src/motion/functionality/layout.ts

```diff
@@ -0,0 +1,274 @@
+import { ValueAnimationControls } from "../../animation/ValueAnimationControls"
+import { MotionValuesMap } from "../utils/use-motion-values"
+import { SyncLayoutContext } from "../../components/SyncLayout"
+import { layoutSync, useLayoutSync } from "../../utils/use-layout-sync"
+import { TargetAndTransition, Transition } from "../../types"
+
+interface Layout {
+    left: number
+    top: number
+    width: number
+    height: number
+}
+
+// We measure the positional delta as x/y as we're actually going to figure out
+// and track the motion of the component's visual center.
+interface LayoutDelta {
+    x: number
+    y: number
+    width: number
+    height: number
+}
+
+// We use both offset and bounding box measurements, and they need to be handled slightly differently
+interface LayoutType {
+    measure: (element: HTMLElement) => Layout
+    getLayout: (info: VisualInfo) => Layout
+}
+
+interface VisualInfo {
+    offset: Layout
+    boundingBox: Layout
+}
+
+const defaultLayoutTransition = {
+    duration: 0.8,
+    ease: [0.45, 0.05, 0.19, 1.0],
+}
+
+function isHTMLElement(
+    element?: Element | HTMLElement | null
+): element is HTMLElement {
+    return element instanceof HTMLElement
+}
+
+function isResolver(
+    transition: AnimationProps["layoutTransition"]
+): transition is ResolveLayoutTransition {
+    return typeof transition === "function"
+}
+
+// Find the center of a Layout definition. We do this to account for potential changes
+// in the top/left etc that are actually just as a result of width/height changes.
+function findCenter({ top, left, width, height }: Layout) {
+    const right = left + width
+    const bottom = top + height
+    return {
+        x: (left + right) / 2,
+        y: (top + bottom) / 2,
+    }
+}
+
+function measureDelta(prev: Layout, next: Layout): LayoutDelta {
+    const prevCenter = findCenter(prev)
+    const nextCenter = findCenter(next)
+
+    return {
+        x: prevCenter.x - nextCenter.x,
+        y: prevCenter.y - nextCenter.y,
+        width: prev.width - next.width,
+        height: prev.height - next.height,
+    }
+}
+
+const offset: LayoutType = {
+    getLayout: ({ offset }) => offset,
+    measure: element => {
+        const { offsetLeft, offsetTop, offsetWidth, offsetHeight } = element
+
+        return {
+            left: offsetLeft,
+            top: offsetTop,
+            width: offsetWidth,
+            height: offsetHeight,
+        }
+    },
+}
+const boundingBox: LayoutType = {
+    getLayout: ({ boundingBox }) => boundingBox,
+    measure: element => {
+        const { left, top, width, height } = element.getBoundingClientRect()
+        return { left, top, width, height }
+    },
+}
+
+function readPositionStyle(element: HTMLElement) {
+    return window.getComputedStyle(element).position
+}
+
+function getLayoutType(prev: string | null, next: string | null): LayoutType {
+    return prev === next ? offset : boundingBox
```

Comment:
I don't understand why this difference exists? Why do we use offset when they are the same?

Reply:
They aren't the same.`offset` is calmer because it doesn't account for transforms, so for instance if we always used bounding box it takes more work to figure out if an element has changed size. But it's also not a reliable measure when switching between `position` types - but in these instances we can also presume that size has changed so it's ok to use bounding box.